### PR TITLE
Implement strings in adapter modules

### LINKED ISSUE #4309: Implement support for utf16+latin1 in the component model
Currently support for utf16+latin1 is not implemented in Wasmtime, but we'll need to finish this and test it before the component model is considered done.

In general I'd expect that this would use the [`encoding_rs` crate](https://crates.io/crates/encoding_rs) for the internal details of latin1 to avoid open-coding that in Wasmtime itself.

## Lowering

[Lowering a string into wasm](https://github.com/bytecodealliance/wasmtime/blob/87007c58393813f0257f710a92006024c4909a67/crates/wasmtime/src/component/func/typed.rs#L1075-L1077) is currently unimplemented. I think that this is required to implement the `store_string_to_latin1_or_utf16` function in the [canonical ABI explainer](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md). My current understanding is that even if we could implement something more optimal in Rust we can't do that because the semantics of lowering are already specified.

I believe the pseudo-code there does most of the fiddly bits but some small helpers in `encoding_rs` are probably going to be required.

## Lifting 

[Calculation of the byte length](https://github.com/bytecodealliance/wasmtime/blob/87007c58393813f0257f710a92006024c4909a67/crates/wasmtime/src/component/func/typed.rs#L1104) and [actually getting the string](https://github.com/bytecodealliance/wasmtime/blob/87007c58393813f0257f710a92006024c4909a67/crates/wasmtime/src/component/func/typed.rs#L1146) are unimplemented. I think that we're free to use `encoding_rs` here however we see fit. Probably the [`decode_latin1`](https://docs.rs/encoding_rs/latest/encoding_rs/mem/fn.decode_latin1.html) function will be useful here. 

## Other notes

I am personally unfamilar with `latin1` as an encoding. I don't know if an arbitrary list of types are guaranteed to be valid latin1 or not. (the infallibility of [`decode_latin1`](https://docs.rs/encoding_rs/latest/encoding_rs/mem/fn.decode_latin1.html) seems odd to me). 

Using `encoding_rs` may be a better option for utf16 decoding we currently do (and maybe even utf8 since `encoding_rs` can probably do simd things that the standard library can't). If someone's intrepid it might be interesting to try to benchmark this and see if it's beneficial to use `encoding_rs` for almost everything.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is a hefty addition to Wasmtime's support for the component
model. This implements the final remaining type (in the current type
hierarchy) unimplemented in adapter module trampolines: strings. Strings
are the most complicated type to implement in adapter trampolines
because they are highly structured chunks of data in memory (according
to specific encodings). Additionally each lift/lower operation can
choose its own encoding for strings meaning that Wasmtime, the host, may
have to convert between any pairwise ordering of string encodings.

The `CanonicalABI.md` in the component-model repo in general specifies
all the fiddly bits of string encoding so there's not a ton of wiggle
room for Wasmtime to get creative. This PR largely "just" implements
that. The high-level architecture of this implementation is:

* Fused adapters are first identified to determine src/dst string
  encodings. This statically fixes what transcoding operation is being
  performed.

* The generated adapter will be responsible for managing calls to
  `realloc` and performing bounds checks. The adapter itself does not
  perform memory copies or validation of string contents, however.
  Instead each transcoding operation is modeled as an imported function
  into the adapter module.  This means that the adapter module
  dynamically, during compile time, determines what string transcoders
  are needed. Note that an imported transcoder is not only parameterized
  over the transcoding operation but additionally which memory is the
  source and which is the destination.

* The imported core wasm functions are modeled as a new
  `CoreDef::Transcoder` structure. These transcoders end up being small
  Cranelift-compiled trampolines. The Cranelift-compiled trampoline will
  load the actual base pointer of memory and add it to the relative
  pointers passed as function arguments. This trampoline then calls a
  transcoder "libcall" which enters Rust-defined functions for actual
  transcoding operations.

* Each possible transcoding operation is implemented in Rust with a
  unique name and a unique signature depending on the needs of the
  transcoder. I've tried to document inline what each transcoder does.

This means that the `Module::translate_string` in adapter modules is by
far the largest translation method. The main reason for this is due to
the management around calling the imported transcoder functions in the
face of validating string pointer/lengths and performing the dance of
`realloc`-vs-transcode at the right time. I've tried to ensure that each
individual case in transcoding is documented well enough to understand
what's going on as well.

Additionally in this PR is a full implementation in the host for the
`latin1+utf16` encoding which means that both lifting and lowering host
strings now works with this encoding.

Currently the implementation of each transcoder function is likely far
from optimal. Where possible I've leaned on the standard library itself
and for latin1-related things I'm leaning on the `encoding_rs` crate. I
initially tried to implement everything with `encoding_rs` but was
unable to uniformly do so easily. For now I settled on trying to get a
known-correct (even in the face of endianness) implementation for all of
these transcoders. If an when performance becomes an issue it should be
possible to implement more optimized versions of each of these
transcoding operations.

Testing this commit has been somewhat difficult and my general plan,
like with the `(list T)` type, is to rely heavily on fuzzing to cover
the various cases here. In this PR though I've added a simple test that
pushes some statically known strings through all the pairs of encodings
between source and destination. I've attempted to pick "interesting"
strings that one way or another stress the various paths in each
transcoding operation to ideally get full branch coverage there.
Additionally a suite of "negative" tests have also been added to ensure
that validity of encoding is actually checked.

Closes #4309

### CODE REVIEW

**Comment by alexcrichton** [comment]
I also realize that this is an extremely large PR unfortunately. I've tried to be thorough with tests but I have almost certainly forgot something here and there. My next plan is to enhance the existing component model fuzzer to have an adapter module generated as well which should exercise more successful cases, but I don't know how to more systematically test the "negative" cases.

**Review by fitzgen — APPROVED** [review]

File: crates/environ/src/fact/trampoline.rs [hunk 1]
```diff
@@ -636,6 +636,689 @@ impl Compiler<'_, '_> {
...
+        // reallocations. First a buffer of size input-code-units is used to try
+        // to get the transcoding correct on the first try. If that fails the
+        // maximum worst-case size is used and then that is resized down if it's
+        // too large.
+        //
+        // The host transcoding function imported here will receive src ptr/len
+        // and dst ptr/len and return how many code units were consumed on both
+        // sides. The amount of code units consumed in the source dictates which
+        // branches are taken in this conversion.
+        let deflate_to_utf8 = |me: &mut Self, src: FE| {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/environ/src/fact/trampoline.rs]: Sure, I've split out everything now.

File: crates/environ/src/fact/trampoline.rs [hunk 2]
```diff
@@ -636,6 +636,689 @@ impl Compiler<'_, '_> {
...
+            validate_string_length(me, src);
+
+            // Optimistically assume that the code unit length of the source is
+            // all that's needed in the destination. Perform that allocaiton
+            // here and proceed to transcoding below.
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/environ/src/fact/trampoline.rs]: Indeed!

File: crates/environ/src/fact/trampoline.rs [hunk 3]
```diff
@@ -636,6 +636,689 @@ impl Compiler<'_, '_> {
...
+            me.convert_src_len_to_dst(src_len, src_opts.ptr(), dst_opts.ptr());
+            me.instruction(LocalTee(dst_len));
+            me.instruction(LocalSet(dst_byte_len));
+            let dst_mem = me.malloc(dst_opts, MallocSize::Local(dst_byte_len), 1);
+
+            // If the string is non-empty encoding is attempted.
+            me.instruction(LocalGet(src_len));
+            me.ptr_if(src_opts, BlockType::Empty);
+
+            // Ensure buffers are all in-bounds
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/environ/src/fact/trampoline.rs]: This was written before the spec changed, and I was originally going to land this first and then fixup all our handling of zero-length arrays and things but I went ahead and updated this.
